File: packages/core/src/routes/perspectives.ts
/**
 * Perspective routes — thin HTTP layer over PerspectiveService.
 */

import { Hono } from 'hono'
import type { Services } from '../services'
import { mapStatusCode } from './server'

export function perspectiveRoutes(services: Services) {
  const app = new Hono()

  app.post('/perspectives', async (c) => {
    const body = await c.req.json<{
      bullet_id: string
      content: string
      target_archetype?: string
      domain?: string
      framing?: string
      auto_approve?: boolean
    }>()
    if (!body.bullet_id) {
      return c.json({ error: { code: 'VALIDATION_ERROR', message: 'bullet_id is required' } }, 400)
    }
    if (!body.content) {
      return c.json({ error: { code: 'VALIDATION_ERROR', message: 'content is required' } }, 400)
    }
    const result = await services.perspectives.createPerspective(body)
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: result.data }, 201)
  })

  app.get('/perspectives', async (c) => {
    const offset = Math.max(0, parseInt(c.req.query('offset') ?? '0', 10) || 0)
    const limit = Math.min(200, Math.max(1, parseInt(c.req.query('limit') ?? '50', 10) || 50))
    const filter: Record<string, string> = {}
    if (c.req.query('bullet_id')) filter.bullet_id = c.req.query('bullet_id')!
    if (c.req.query('target_archetype')) filter.target_archetype = c.req.query('target_archetype')!
    if (c.req.query('domain')) filter.domain = c.req.query('domain')!
    if (c.req.query('framing')) filter.framing = c.req.query('framing')!
    if (c.req.query('status')) filter.status = c.req.query('status')!
    if (c.req.query('source_id')) filter.source_id = c.req.query('source_id')!
    if (c.req.query('search')) filter.search = c.req.query('search')!
    if (c.req.query('archetype')) filter.target_archetype = c.req.query('archetype')!

    const result = await services.perspectives.listPerspectives(filter, offset, limit)
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: result.data, pagination: result.pagination })
  })

  app.get('/perspectives/:id', async (c) => {
    const result = await services.perspectives.getPerspectiveWithChain(c.req.param('id'))
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: result.data })
  })

  app.patch('/perspectives/:id', async (c) => {
    const body = await c.req.json()
    const result = await services.perspectives.updatePerspective(c.req.param('id'), body)
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: result.data })
  })

  app.delete('/perspectives/:id', async (c) => {
    const result = await services.perspectives.deletePerspective(c.req.param('id'))
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.body(null, 204)
  })

  app.patch('/perspectives/:id/approve', async (c) => {
    const result = await services.perspectives.approvePerspective(c.req.param('id'))
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: result.data })
  })

  app.patch('/perspectives/:id/reject', async (c) => {
    const body = await c.req.json<{ rejection_reason?: string }>()
    const result = await services.perspectives.rejectPerspective(c.req.param('id'), body.rejection_reason ?? '')
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: result.data })
  })

  app.patch('/perspectives/:id/reopen', async (c) => {
    const result = await services.perspectives.reopenPerspective(c.req.param('id'))
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: result.data })
  })

  return app
}

File: crates/forge-server/src/routes/perspectives.rs
//! Perspective CRUD + status transition routes.
//!
//! Mirrors `packages/core/src/routes/perspectives.ts`.

use axum::extract::{Path, Query, State};
use axum::routing::{get, patch, post};
use axum::{Json, Router};
use serde::Deserialize;

use forge_core::{
    CreatePerspectiveInput, Framing, PaginationParams, Perspective, PerspectiveFilter,
    PerspectiveStatus, PerspectiveWithChain, UpdatePerspectiveInput,
};
use forge_sdk::db::PerspectiveStore;

use crate::db::with_conn;
use crate::error::ApiError;
use crate::response::{ApiData, ApiList, Created, NoContent};
use crate::state::SharedState;

// ── Query params ────────────────────────────────────────────────────

#[derive(Debug, Deserialize, Default)]
pub struct PerspectiveListQuery {
    pub offset: Option<i64>,
    pub limit: Option<i64>,
    pub bullet_id: Option<String>,
    pub target_archetype: Option<String>,
    pub archetype: Option<String>,
    pub domain: Option<String>,
    pub framing: Option<String>,
    pub status: Option<String>,
    pub source_id: Option<String>,
    pub search: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct RejectBody {
    pub rejection_reason: Option<String>,
}

// ── Handlers ────────────────────────────────────────────────────────

async fn create_perspective(
    State(state): State<SharedState>,
    Json(input): Json<CreatePerspectiveInput>,
) -> Result<Created<Perspective>, ApiError> {
    let result =
        with_conn(&state, move |conn| PerspectiveStore::create(conn, &input)).await?;
    Ok(Created(result))
}

async fn list_perspectives(
    State(state): State<SharedState>,
    Query(q): Query<PerspectiveListQuery>,
) -> Result<Json<ApiList<Perspective>>, ApiError> {
    let filter = PerspectiveFilter {
        bullet_id: q.bullet_id,
        target_archetype: q.target_archetype.or(q.archetype),
        domain: q.domain,
        framing: q.framing.and_then(|s| s.parse::<Framing>().ok()),
        status: q.status.and_then(|s| s.parse::<PerspectiveStatus>().ok()),
        source_id: q.source_id,
        search: q.search,
    };
    let pg = PaginationParams {
        offset: Some(q.offset.unwrap_or(0).max(0)),
        limit: Some(q.limit.unwrap_or(50).clamp(1, 200)),
    };

    let (data, pagination) =
        with_conn(&state, move |conn| PerspectiveStore::list(conn, &filter, &pg)).await?;

    Ok(Json(ApiList { data, pagination }))
}

async fn get_perspective(
    State(state): State<SharedState>,
    Path(id): Path<String>,
) -> Result<Json<ApiData<PerspectiveWithChain>>, ApiError> {
    let result = with_conn(&state, move |conn| {
        PerspectiveStore::get_with_chain(conn, &id)?.ok_or_else(|| {
            forge_core::ForgeError::NotFound {
                entity_type: "Perspective".into(),
                id: id.clone(),
            }
        })
    })
    .await?;
    Ok(Json(ApiData { data: result }))
}

async fn update_perspective(
    State(state): State<SharedState>,
    Path(id): Path<String>,
    Json(input): Json<UpdatePerspectiveInput>,
) -> Result<Json<ApiData<Perspective>>, ApiError> {
    let result =
        with_conn(&state, move |conn| PerspectiveStore::update(conn, &id, &input)).await?;
    Ok(Json(ApiData { data: result }))
}

async fn delete_perspective(
    State(state): State<SharedState>,
    Path(id): Path<String>,
) -> Result<NoContent, ApiError> {
    with_conn(&state, move |conn| PerspectiveStore::delete(conn, &id)).await?;
    Ok(NoContent)
}

async fn approve_perspective(
    State(state): State<SharedState>,
    Path(id): Path<String>,
) -> Result<Json<ApiData<Perspective>>, ApiError> {
    let result = with_conn(&state, move |conn| {
        PerspectiveStore::transition_status(conn, &id, PerspectiveStatus::Approved, None)
    })
    .await?;
    Ok(Json(ApiData { data: result }))
}

async fn reject_perspective(
    State(state): State<SharedState>,
    Path(id): Path<String>,
    Json(body): Json<RejectBody>,
) -> Result<Json<ApiData<Perspective>>, ApiError> {
    let reason = body.rejection_reason.unwrap_or_default();
    let result = with_conn(&state, move |conn| {
        PerspectiveStore::transition_status(
            conn,
            &id,
            PerspectiveStatus::Rejected,
            Some(&reason),
        )
    })
    .await?;
    Ok(Json(ApiData { data: result }))
}

async fn reopen_perspective(
    State(state): State<SharedState>,
    Path(id): Path<String>,
) -> Result<Json<ApiData<Perspective>>, ApiError> {
    let result = with_conn(&state, move |conn| {
        PerspectiveStore::transition_status(conn, &id, PerspectiveStatus::Draft, None)
    })
    .await?;
    Ok(Json(ApiData { data: result }))
}

// ── Router ──────────────────────────────────────────────────────────

pub fn router() -> Router<SharedState> {
    Router::new()
        .route("/perspectives", post(create_perspective).get(list_perspectives))
        .route(
            "/perspectives/{id}",
            get(get_perspective)
                .patch(update_perspective)
                .delete(delete_perspective),
        )
        .route("/perspectives/{id}/approve", patch(approve_perspective))
        .route("/perspectives/{id}/reject", patch(reject_perspective))
        .route("/perspectives/{id}/reopen", patch(reopen_perspective))
}
